.github/workflows/windows.yml

```diff
@@ -42,52 +42,204 @@ env:
+            echo 'is_self_hosted=false' | tee -a $GITHUB_OUTPUT
+            exit 0
+          }
+
+          # Generate a unique id that allows the workload job to find the runner
+          # we are reserving for it (via runner labels), and allows the timeout
+          # job to find the workload job run (via the job’s friendly name), even
+          # if there are multiple instances in the workflow call tree.
+          random32=$(LC_ALL=C < /dev/urandom tr -dC 0-9a-f 2> /dev/null | head -c 8)
+          unique_id=${{ github.repository }}/${{ github.run_id }}/$random32
+          echo "unique_id=$unique_id" | tee -a $GITHUB_OUTPUT
+
+          # Disable self-hosted runners by creating a repository variable named
+          # NO_SELF_HOSTED_RUNNERS with any non-empty value.
+          # <https://github.com/servo/servo/settings/variables/actions>
+          if [ -n '${{ vars.NO_SELF_HOSTED_RUNNERS }}' ]; then
+            echo 'NO_SELF_HOSTED_RUNNERS is set!'
+            fall_back_to_github_hosted
+          fi
+
+          # RUNNER_API_TOKEN secret will be unavailable in forks.
+          if [ -z "$GH_TOKEN" ]; then
+            echo 'RUNNER_API_TOKEN not set!'
+            fall_back_to_github_hosted
+          fi
+
+          runners=$(mktemp)
+          gh api "$self_hosted_runner_scope" > $runners
+
+          # Find a runner that is online, not busy, and not already reserved for
+          # any job (label prefix “reserved-for:”).
+          runner_id=$(mktemp)
+          if ! < $runners > $runner_id jq \
+            --arg self_hosted_image_label "$self_hosted_image_label" -e '
+              .runners
+              | map(select(.status == "online" and .busy == false))
+              | map(select([.labels[].name] | index($self_hosted_image_label) | not | not))
+              | map(select([.labels[].name] | map(startswith("reserved-for:")) | index(true) | not))
+              | first | .id'; then
+            echo 'No self-hosted runners available!'
+            fall_back_to_github_hosted
+          fi
+
+          # Reserve that runner by adding a label containing the unique id.
+          # Job concurrency ensures that runners never get assigned twice.
+          reserved_since=$(date +\%s)
+          gh api "$self_hosted_runner_scope/$(cat $runner_id)/labels" \
+            -f "labels[]=reserved-for:$unique_id" \
+            -f "labels[]=reserved-since:$reserved_since" \
+            --method POST --silent
+          echo "selected_runner_label=reserved-for:$unique_id" | tee -a $GITHUB_OUTPUT
+          echo 'is_self_hosted=true' | tee -a $GITHUB_OUTPUT
+
+  # In the unlikely event a self-hosted runner was selected and reserved but it
+  # goes down before the workload starts, cancel the workflow run.
+  runner-timeout:
+    needs:
+      - runner-select
+    if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) }}
+    name: Runner timeout
```

Comment:
```suggestion
    name: Detect Runner Timeout
```

Reply:
done in bcc52e09aefc7b32424833b414e338cc5f772586